File: statsPy/core/10_correlation_and_dependence.py
# ---
# jupyter:
#   jupytext:
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#   kernelspec:
#     display_name: Python 3
#     language: python
#     name: python3
# ---

# %% [markdown]
# # Module 10: Correlation and dependence
#
# **Curriculum link:** `stats.md` -> Topic 10, equations (10.1)-(10.10)
#
# ## What you will learn
#
# 1. Pearson (10.2) measures **linear** association only, and is not invariant
#    to the log transforms omics uses everywhere.
# 2. Fisher's $z$ (10.3) for correct intervals.
# 3. Partial correlation (10.5) and the precision matrix (10.6): marginal vs
#    conditional independence.
# 4. Simpson's paradox and repeated-measures correlation (10.7).
# 5. Correlation is **not agreement**: Bland-Altman (10.8), Lin's CCC (10.9).
# 6. Correlation matrices under multiplicity, and shrinkage (10.10).

# %%
import os

import numpy as np
import pandas as pd
import scipy.stats as st
import statsmodels.formula.api as smf
import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt

MODULE_NAME = "10_correlation_and_dependence"
OUT = os.path.join(os.environ.get("STATS_OUT", "results"), MODULE_NAME)
os.makedirs(OUT, exist_ok=True)


def header(t):
    print("\n" + "=" * 72); print(t); print("=" * 72)


# %% [markdown]
# ## 1. What Pearson does and does not detect

# %%
header("1. r = 0 does not mean independent (10.1)-(10.2)")

rng = np.random.default_rng(1001)
n = 2000
x = rng.uniform(-3, 3, n)
cases = {
    "linear":        (x, 1.5 * x + rng.normal(0, 1, n)),
    "quadratic":     (x, x**2 + rng.normal(0, 1, n)),
    "monotone (exp)": (x, np.exp(x) + rng.normal(0, 1, n)),
    "independent":   (x, rng.normal(0, 1, n)),
}
print(f"  {'relationship':<18}{'Pearson r':>11}{'Spearman':>10}{'Kendall':>10}")
for name, (a, b) in cases.items():
    print(f"  {name:<18}{st.pearsonr(a, b).statistic:>11.3f}"
          f"{st.spearmanr(a, b).statistic:>10.3f}"
          f"{st.kendalltau(a, b).statistic:>10.3f}")
print("\n  'quadratic': r ~ 0 despite a perfect deterministic relationship.")
print("  'monotone' : Spearman/Kendall detect it far better than Pearson.")

# Scale dependence: critical for omics, where log is everywhere.
counts_a = rng.lognormal(3, 1, 500)
counts_b = counts_a ** 1.3 * rng.lognormal(0, 0.4, 500)
print(f"\n  Same data, different scale:")
print(f"    r on raw scale      = {st.pearsonr(counts_a, counts_b).statistic:.3f}")
print(f"    r on log scale      = "
      f"{st.pearsonr(np.log(counts_a), np.log(counts_b)).statistic:.3f}")
print(f"    Spearman (invariant)= {st.spearmanr(counts_a, counts_b).statistic:.3f}")
print("  ALWAYS state the scale on which a correlation was computed.")

# %% [markdown]
# ## 2. Fisher's z-transform, eq. (10.3)
#
# $$z=\operatorname{artanh}(r)\ \dot\sim\ \mathcal N\!\Big(\operatorname{artanh}(\rho),\tfrac1{n-3}\Big)$$

# %%
header("2. Correct confidence intervals for a correlation (10.3)")

def fisher_ci(r, n, alpha=0.05):
    """stats.md eq. (10.3): CI built on the z scale, mapped back with tanh."""
    z = np.arctanh(r)
    se = 1.0 / np.sqrt(n - 3)
    zc = st.norm.ppf(1 - alpha / 2)
    return np.tanh([z - zc * se, z + zc * se])


for r, n_ in [(0.9, 10), (0.9, 100), (0.3, 20), (-0.6, 15)]:
    lo, hi = fisher_ci(r, n_)
    naive = np.array([r - 1.96 / np.sqrt(n_), r + 1.96 / np.sqrt(n_)])
    print(f"  r={r:+.2f}, n={n_:>4}: Fisher CI [{lo:+.3f}, {hi:+.3f}]   "
          f"naive symmetric [{naive[0]:+.3f}, {naive[1]:+.3f}]")
print("\n  The Fisher interval is asymmetric and stays inside [-1,1]; the naive")
print("  one escapes the parameter space. Coverage check:")

rng = np.random.default_rng(1002)
RHO, n_ = 0.7, 15
cov_f = cov_n = 0
for _ in range(20000):
    xy = rng.multivariate_normal([0, 0], [[1, RHO], [RHO, 1]], n_)
    r = np.corrcoef(xy[:, 0], xy[:, 1])[0, 1]
    lo, hi = fisher_ci(r, n_)
    cov_f += lo <= RHO <= hi
    cov_n += (r - 1.96/np.sqrt(n_)) <= RHO <= (r + 1.96/np.sqrt(n_))
print(f"    Fisher coverage = {cov_f/20000:.1%}   naive coverage = "
      f"{cov_n/20000:.1%}   (nominal 95%)")

# %% [markdown]
# ## 3. Partial correlation and the precision matrix, eq. (10.5)-(10.6)
#
# A zero in the **covariance** matrix means *marginal* independence; a zero in
# the **precision** matrix $\boldsymbol\Sigma^{-1}$ means *conditional*
# independence. Very different claims.

# %%
header("3. Marginal vs conditional independence (10.5)-(10.6)")

rng = np.random.default_rng(1003)
n = 3000
# Chain structure: Z -> X and Z -> Y. X and Y are marginally correlated but
# CONDITIONALLY INDEPENDENT given Z. This is the classic confounding pattern.
z = rng.normal(0, 1, n)
x = 1.2 * z + rng.normal(0, 1, n)
y = 1.5 * z + rng.normal(0, 1, n)


def partial_corr(a, b, c):
    """stats.md eq. (10.5)."""
    rab, rac, rbc = (st.pearsonr(a, b).statistic, st.pearsonr(a, c).statistic,
                     st.pearsonr(b, c).statistic)
    return (rab - rac * rbc) / np.sqrt((1 - rac**2) * (1 - rbc**2))


print(f"  marginal r(X,Y)        = {st.pearsonr(x, y).statistic:+.4f}")
print(f"  partial  r(X,Y | Z)    = {partial_corr(x, y, z):+.4f}  (eq. 10.5)")

# The precision-matrix route (eq. 10.6) gives the same answer.
M = np.column_stack([x, y, z])
S = np.cov(M, rowvar=False)
Omega = np.linalg.inv(S)
pc_from_precision = -Omega[0, 1] / np.sqrt(Omega[0, 0] * Omega[1, 1])
print(f"  from precision matrix  = {pc_from_precision:+.4f}  (eq. 10.6) <- same")
print("\n  X and Y look strongly associated but are conditionally independent")
print("  given Z. A co-expression network built on MARGINAL correlations fills")
print("  with such indirect edges (stats.md Topic 30).")

# %% [markdown]
# ## 4. Simpson's paradox and repeated measures, eq. (10.7)

# %%
header("4. Pooling repeated measures can REVERSE the association (10.7)")

rng = np.random.default_rng(1004)
n_subj, n_visits = 12, 8
records = []
for s in range(n_subj):
    # Subjects with high baseline biomarker tend to have LOW outcome (between-
    # subject effect), but WITHIN each subject the relationship is POSITIVE.
    base_x = rng.normal(s * 1.0, 0.3)
    base_y = 30 - 2.0 * s + rng.normal(0, 0.8)
    for v in range(n_visits):
        xv = base_x + rng.normal(0, 0.5)
        yv = base_y + 1.5 * (xv - base_x) + rng.normal(0, 0.4)
        records.append({"subject": f"S{s:02d}", "x": xv, "y": yv})
d = pd.DataFrame(records)

pooled_r = st.pearsonr(d["x"], d["y"]).statistic
# Repeated-measures correlation (eq. 10.7): common within-subject slope.
fit = smf.ols("y ~ x + C(subject)", data=d).fit()
slope = fit.params["x"]
# Sign * sqrt(partial R^2) for the x term.
fit0 = smf.ols("y ~ C(subject)", data=d).fit()
ss_model = fit0.ssr - fit.ssr
r_rm = np.sign(slope) * np.sqrt(ss_model / (ss_model + fit.ssr))

print(f"  naive pooled correlation      r = {pooled_r:+.4f}")
print(f"  repeated-measures correlation r = {r_rm:+.4f}   (eq. 10.7)")
print(f"  common within-subject slope     = {slope:+.4f}")
print("\n  The pooled correlation has the OPPOSITE SIGN to the within-subject")
print("  relationship. Handle donor structure before interpreting correlation.")

# %% [markdown]
# ## 5. Correlation is not agreement, eq. (10.8)-(10.9)

# %%
header("5. Bland-Altman and Lin's CCC (10.8)-(10.9)")

rng = np.random.default_rng(1005)
truth = rng.normal(100, 15, 120)
assay1 = truth + rng.normal(0, 2, 120)
assay2 = 1.9 * truth + 8 + rng.normal(0, 2, 120)    # perfectly correlated, wrong scale

def lins_ccc(a, b):
    """stats.md eq. (10.9)."""
    va, vb = a.var(ddof=0), b.var(ddof=0)
    cov = np.cov(a, b, ddof=0)[0, 1]
    return 2 * cov / (va + vb + (a.mean() - b.mean()) ** 2)


def bland_altman(a, b):
    """stats.md eq. (10.8): bias and 95% limits of agreement."""
    diff = a - b
    return {"bias": diff.mean(),
            "loa": (diff.mean() - 1.96 * diff.std(ddof=1),
                    diff.mean() + 1.96 * diff.std(ddof=1))}


print(f"  Pearson r(assay1, assay2) = {st.pearsonr(assay1, assay2).statistic:.5f}"
      f"   <- 'excellent agreement'?")
print(f"  Lin's CCC                 = {lins_ccc(assay1, assay2):.5f}"
      f"   <- no, they disagree badly")
ba = bland_altman(assay1, assay2)
print(f"  Bland-Altman bias         = {ba['bias']:+.2f}")
print(f"  95% limits of agreement   = [{ba['loa'][0]:+.2f}, {ba['loa'][1]:+.2f}]")
print("\n  r near 1 with CCC near 0: the assays rank samples identically but")
print("  report entirely different numbers. For method comparison always use")
print("  (10.8)-(10.9), never r.")

# %% [markdown]
# ## 6. Correlation matrices: multiplicity and shrinkage, eq. (10.10)

# %%
header("6. Co-expression under multiplicity, and shrinkage (10.10)")

rng = np.random.default_rng(1006)
G, n_samp = 200, 20
# COMPLETELY independent features: every correlation we find is noise.
X = rng.normal(0, 1, (n_samp, G))
R = np.corrcoef(X, rowvar=False)
iu = np.triu_indices(G, 1)
rs = R[iu]
# t-test for each correlation: t = r sqrt((n-2)/(1-r^2)), df = n-2
tvals = rs * np.sqrt((n_samp - 2) / (1 - rs**2))
pvals = 2 * st.t.sf(np.abs(tvals), n_samp - 2)

print(f"  G = {G} features, n = {n_samp} samples, ALL truly independent")
print(f"  number of pairwise tests = G(G-1)/2 = {len(rs):,}")
print(f"  'significant' at p < 0.05 (unadjusted): {np.sum(pvals < 0.05):,} "
      f"({np.mean(pvals < 0.05):.1%})")
print(f"  |r| > 0.5 purely by chance             : {np.sum(np.abs(rs) > 0.5):,}")
from statsmodels.stats.multitest import multipletests
print(f"  surviving BH at 0.05                   : "
      f"{np.sum(multipletests(pvals, method='fdr_bh')[0]):,}")
print("\n  Co-expression networks built from unadjusted correlations are mostly")
print("  noise, especially at the small n typical of omics.")

def ledoit_wolf_shrinkage(X):
    """stats.md eq. (10.10): shrink the sample covariance to a scaled identity."""
    n, p = X.shape
    Xc = X - X.mean(axis=0)
    S = Xc.T @ Xc / n
    mu = np.trace(S) / p
    T = mu * np.eye(p)                              # structured target
    d2 = np.sum((S - T) ** 2) / p
    b2_bar = np.mean([np.sum((np.outer(Xc[i], Xc[i]) - S) ** 2) / p
                      for i in range(n)]) / n
    b2 = min(b2_bar, d2)
    lam = b2 / d2
    return (1 - lam) * S + lam * T, lam


S_shrunk, lam = ledoit_wolf_shrinkage(X)
S_raw = np.cov(X, rowvar=False)
print(f"\n  p={G} > n={n_samp}, so the sample covariance is SINGULAR:")
print(f"    rank(S_sample) = {np.linalg.matrix_rank(S_raw)} (needs {G})")
print(f"    Ledoit-Wolf shrinkage intensity lambda = {lam:.4f}")
print(f"    rank(S_shrunk) = {np.linalg.matrix_rank(S_shrunk)}  -> invertible,")
print("    which is what makes the precision matrix of eq. (10.6) computable.")

# %% [markdown]
# ## 7. Figure

# %%
fig, axes = plt.subplots(1, 4, figsize=(17, 4))

a, b = cases["quadratic"]
axes[0].scatter(a, b, s=3, alpha=0.3)
axes[0].set_title(f"r = {st.pearsonr(a,b).statistic:.3f}, yet deterministic",
                  fontsize=9)

for s, grp in d.groupby("subject"):
    axes[1].scatter(grp["x"], grp["y"], s=12)
xs = np.linspace(d["x"].min(), d["x"].max(), 10)
axes[1].plot(xs, np.polyval(np.polyfit(d["x"], d["y"], 1), xs), "k--", lw=2,
             label=f"pooled r={pooled_r:+.2f}")
axes[1].set_title("Eq. (10.7): Simpson's paradox", fontsize=9)
axes[1].legend(fontsize=7)

diff = assay1 - assay2
avg = (assay1 + assay2) / 2
axes[2].scatter(avg, diff, s=10)
axes[2].axhline(ba["bias"], color="k")
for v in ba["loa"]:
    axes[2].axhline(v, color="red", ls="--")
axes[2].set_xlabel("mean of the two assays"); axes[2].set_ylabel("difference")
axes[2].set_title("Eq. (10.8): Bland-Altman", fontsize=9)

axes[3].hist(rs, bins=60, color="steelblue")
axes[3].axvline(0.5, color="red", ls="--"); axes[3].axvline(-0.5, color="red", ls="--")
axes[3].set_xlabel("pairwise r among INDEPENDENT features")
axes[3].set_title(f"n={n_samp}: |r|>0.5 happens {np.mean(np.abs(rs)>0.5):.1%} "
                  f"of the time", fontsize=8)

fig.tight_layout()
fig.savefig(os.path.join(OUT, "correlation.png"), dpi=110, bbox_inches="tight")
plt.close(fig)
print(f"\nFigure written to {OUT}/correlation.png")

# %% [markdown]
# ## Decision rules (from `stats.md` Topic 10)
#
# 1. Pearson for linear association on a justified scale; Spearman/Kendall for
#    monotone association or ordinal data.
# 2. Handle donor structure explicitly (10.7) before interpreting correlation
#    across repeated measurements.
# 3. Use Fisher's $z$ (10.3) for intervals and tests.
# 4. For method comparison, report (10.8)-(10.9), not $r$.
# 5. Apply log-ratio transformation before correlating compositions (Topic 26).
#
# **Next:** `11_linear_models.py`

File: statsR/core/10_correlation_and_dependence.R
#' ---
#' title: "Module 10 - Correlation and dependence"
#' output: html_document
#' ---
#'
#' **Curriculum link:** `stats.md` -> Topic 10, equations (10.1)-(10.10)
#'
#' ## What you will learn
#'
#' 1. Pearson measures LINEAR association only, and is scale-dependent.
#' 2. Fisher's z (10.3) for correct intervals.
#' 3. Partial correlation (10.5) and the precision matrix (10.6).
#' 4. Simpson's paradox and repeated-measures correlation (10.7).
#' 5. Correlation is NOT agreement -- Bland-Altman (10.8), Lin's CCC (10.9).
#' 6. Correlation matrices under multiplicity, and shrinkage (10.10).

#+ setup, message = FALSE
MODULE_NAME <- "10_correlation_and_dependence"
OUT <- file.path(Sys.getenv("STATS_OUT", unset = "results"), MODULE_NAME)
dir.create(OUT, recursive = TRUE, showWarnings = FALSE)
header <- function(txt) cat("\n", strrep("=", 72), "\n", txt, "\n",
                            strrep("=", 72), "\n", sep = "")

#' ## 1. What Pearson does and does not detect

#+ what-pearson
header("1. r = 0 does not mean independent (10.1)-(10.2)")
set.seed(1001)
## Distinct names (`x1`, not `x`): later sections rebind `x` and `n`, and the
## figure at the bottom still needs these. Shadowing is the commonest cause of
## "it ran yesterday" breakage in analysis scripts.
n1 <- 2000; x1 <- runif(n1, -3, 3)
cases <- list(linear = 1.5 * x1 + rnorm(n1), quadratic = x1^2 + rnorm(n1),
              `monotone (exp)` = exp(x1) + rnorm(n1), independent = rnorm(n1))
cat(sprintf("  %-18s%11s%10s%10s\n", "relationship", "Pearson r", "Spearman", "Kendall"))
for (nm in names(cases))
  cat(sprintf("  %-18s%11.3f%10.3f%10.3f\n", nm, cor(x1, cases[[nm]]),
              cor(x1, cases[[nm]], method = "spearman"),
              cor(x1, cases[[nm]], method = "kendall")))
cat("\n  'quadratic': r ~ 0 despite a perfect deterministic relationship.\n")
cat("  'monotone' : Spearman/Kendall detect it far better than Pearson.\n")

ca <- rlnorm(500, 3, 1); cb <- ca^1.3 * rlnorm(500, 0, 0.4)
cat(sprintf("\n  Same data, different scale:\n"))
cat(sprintf("    r on raw scale       = %.3f\n", cor(ca, cb)))
cat(sprintf("    r on log scale       = %.3f\n", cor(log(ca), log(cb))))
cat(sprintf("    Spearman (invariant) = %.3f\n", cor(ca, cb, method = "spearman")))
cat("  ALWAYS state the scale on which a correlation was computed.\n")

#' ## 2. Fisher's z-transform, eq. (10.3)

#+ fisher-z
header("2. Correct confidence intervals for a correlation (10.3)")
fisher_ci <- function(r, n, alpha = 0.05) {
  z <- atanh(r); se <- 1 / sqrt(n - 3)
  tanh(z + c(-1, 1) * qnorm(1 - alpha/2) * se)
}
for (cfg in list(c(0.9, 10), c(0.9, 100), c(0.3, 20), c(-0.6, 15))) {
  r <- cfg[1]; nn <- cfg[2]
  ci <- fisher_ci(r, nn); naive <- r + c(-1, 1) * 1.96 / sqrt(nn)
  cat(sprintf("  r=%+.2f, n=%4d: Fisher CI [%+.3f, %+.3f]   naive [%+.3f, %+.3f]\n",
              r, nn, ci[1], ci[2], naive[1], naive[2]))
}
cat("\n  The Fisher interval is asymmetric and stays inside [-1,1]; the naive\n")
cat("  one escapes the parameter space. Coverage check:\n")
set.seed(1002)
RHO <- 0.7; nn <- 15
cov <- replicate(20000, {
  s <- matrix(c(1, RHO, RHO, 1), 2)
  xy <- MASS::mvrnorm(nn, c(0, 0), s)
  r <- cor(xy[, 1], xy[, 2])
  ci <- fisher_ci(r, nn)
  c(ci[1] <= RHO && RHO <= ci[2],
    (r - 1.96/sqrt(nn)) <= RHO && RHO <= (r + 1.96/sqrt(nn)))
})
cat(sprintf("    Fisher coverage = %.1f%%   naive coverage = %.1f%%   (nominal 95%%)\n",
            100 * mean(cov[1, ]), 100 * mean(cov[2, ])))
cat(sprintf("    cor.test() CI for comparison: matches Fisher by construction.\n"))

#' ## 3. Partial correlation and the precision matrix, eq. (10.5)-(10.6)

#+ partial
header("3. Marginal vs conditional independence (10.5)-(10.6)")
set.seed(1003)
n <- 3000
z <- rnorm(n); x <- 1.2 * z + rnorm(n); y <- 1.5 * z + rnorm(n)
partial_corr <- function(a, b, c) {                             # eq. (10.5)
  rab <- cor(a, b); rac <- cor(a, c); rbc <- cor(b, c)
  (rab - rac * rbc) / sqrt((1 - rac^2) * (1 - rbc^2))
}
S <- cov(cbind(x, y, z)); Omega <- solve(S)
pc_precision <- -Omega[1, 2] / sqrt(Omega[1, 1] * Omega[2, 2])   # eq. (10.6)
cat(sprintf("  marginal r(X,Y)       = %+.4f\n", cor(x, y)))
cat(sprintf("  partial  r(X,Y | Z)   = %+.4f  (eq. 10.5)\n", partial_corr(x, y, z)))
cat(sprintf("  from precision matrix = %+.4f  (eq. 10.6) <- same\n", pc_precision))
cat("\n  X and Y look strongly associated but are conditionally independent\n")
cat("  given Z. Co-expression networks on MARGINAL correlations fill with\n")
cat("  such indirect edges (stats.md Topic 30).\n")

#' ## 4. Simpson's paradox and repeated measures, eq. (10.7)

#+ simpsons
header("4. Pooling repeated measures can REVERSE the association (10.7)")
set.seed(1004)
n_subj <- 12; n_visits <- 8
recs <- do.call(rbind, lapply(seq_len(n_subj), function(s) {
  base_x <- rnorm(1, s, 0.3)
  base_y <- rnorm(1, 30 - 2 * s, 0.8)
  xv <- base_x + rnorm(n_visits, 0, 0.5)
  data.frame(subject = sprintf("S%02d", s), x = xv,
             y = base_y + 1.5 * (xv - base_x) + rnorm(n_visits, 0, 0.4))
}))
pooled_r <- cor(recs$x, recs$y)
fit <- lm(y ~ x + subject, data = recs); fit0 <- lm(y ~ subject, data = recs)
ss_model <- sum(residuals(fit0)^2) - sum(residuals(fit)^2)
r_rm <- sign(coef(fit)["x"]) * sqrt(ss_model / (ss_model + sum(residuals(fit)^2)))
cat(sprintf("  naive pooled correlation      r = %+.4f\n", pooled_r))
cat(sprintf("  repeated-measures correlation r = %+.4f   (eq. 10.7)\n", r_rm))
cat(sprintf("  common within-subject slope     = %+.4f\n", coef(fit)["x"]))
cat("\n  The pooled correlation has the OPPOSITE SIGN to the within-subject\n")
cat("  relationship. Handle donor structure before interpreting correlation.\n")

#' ## 5. Correlation is not agreement, eq. (10.8)-(10.9)

#+ agreement
header("5. Bland-Altman and Lin's CCC (10.8)-(10.9)")
set.seed(1005)
truth <- rnorm(120, 100, 15)
assay1 <- truth + rnorm(120, 0, 2)
assay2 <- 1.9 * truth + 8 + rnorm(120, 0, 2)    # perfectly correlated, wrong scale
lins_ccc <- function(a, b) {                                     # eq. (10.9)
  va <- var(a) * (length(a)-1)/length(a); vb <- var(b) * (length(b)-1)/length(b)
  cv <- cov(a, b) * (length(a)-1)/length(a)
  2 * cv / (va + vb + (mean(a) - mean(b))^2)
}
d_ba <- assay1 - assay2
cat(sprintf("  Pearson r(assay1, assay2) = %.5f   <- 'excellent agreement'?\n",
            cor(assay1, assay2)))
cat(sprintf("  Lin's CCC                 = %.5f   <- no, they disagree badly\n",
            lins_ccc(assay1, assay2)))
cat(sprintf("  Bland-Altman bias         = %+.2f\n", mean(d_ba)))
cat(sprintf("  95%% limits of agreement   = [%+.2f, %+.2f]   (eq. 10.8)\n",
            mean(d_ba) - 1.96 * sd(d_ba), mean(d_ba) + 1.96 * sd(d_ba)))
cat("\n  r near 1 with CCC near 0: the assays RANK samples identically but\n")
cat("  report entirely different numbers. For method comparison always use\n")
cat("  (10.8)-(10.9), never r.\n")

#' ## 6. Correlation matrices: multiplicity and shrinkage, eq. (10.10)

#+ multiplicity-shrinkage
header("6. Co-expression under multiplicity, and shrinkage (10.10)")
set.seed(1006)
G <- 200; n_samp <- 20
X <- matrix(rnorm(n_samp * G), nrow = n_samp)      # COMPLETELY independent
R <- cor(X); iu <- upper.tri(R); rs <- R[iu]
tvals <- rs * sqrt((n_samp - 2) / (1 - rs^2))
pvals <- 2 * pt(abs(tvals), n_samp - 2, lower.tail = FALSE)
cat(sprintf("  G = %d features, n = %d samples, ALL truly independent\n", G, n_samp))
cat(sprintf("  number of pairwise tests = G(G-1)/2 = %s\n",
            format(length(rs), big.mark = ",")))
cat(sprintf("  'significant' at p < 0.05 (unadjusted): %s (%.1f%%)\n",
            format(sum(pvals < 0.05), big.mark = ","), 100 * mean(pvals < 0.05)))
cat(sprintf("  |r| > 0.5 purely by chance             : %s\n",
            format(sum(abs(rs) > 0.5), big.mark = ",")))
cat(sprintf("  surviving BH at 0.05                   : %s\n",
            format(sum(p.adjust(pvals, "BH") < 0.05), big.mark = ",")))
cat("\n  Co-expression networks built from unadjusted correlations are mostly\n")
cat("  noise at the small n typical of omics.\n")

ledoit_wolf <- function(X) {                                     # eq. (10.10)
  n <- nrow(X); p <- ncol(X)
  Xc <- scale(X, center = TRUE, scale = FALSE)
  S <- crossprod(Xc) / n
  mu <- sum(diag(S)) / p
  Tm <- mu * diag(p)
  d2 <- sum((S - Tm)^2) / p
  b2bar <- mean(apply(Xc, 1, function(r) sum((tcrossprod(r) - S)^2) / p)) / n
  lam <- min(b2bar, d2) / d2
  list(S = (1 - lam) * S + lam * Tm, lambda = lam)
}
lw <- ledoit_wolf(X)
cat(sprintf("\n  p=%d > n=%d, so the sample covariance is SINGULAR:\n", G, n_samp))
cat(sprintf("    rank(S_sample) = %d (needs %d)\n", qr(cov(X))$rank, G))
cat(sprintf("    Ledoit-Wolf shrinkage intensity lambda = %.4f\n", lw$lambda))
cat(sprintf("    rank(S_shrunk) = %d  -> invertible, which is what makes the\n",
            qr(lw$S)$rank))
cat("    precision matrix of eq. (10.6) computable.\n")

#' ## 7. Figure

#+ figure
png(file.path(OUT, "correlation.png"), width = 1500, height = 400, res = 110)
par(mfrow = c(1, 4), mar = c(4.2, 4.2, 2.5, 1))
plot(x1, cases$quadratic, pch = ".", col = "steelblue", xlab = "x", ylab = "y",
     main = sprintf("r = %.3f, yet deterministic", cor(x1, cases$quadratic)))
plot(recs$x, recs$y, col = factor(recs$subject), pch = 16, cex = 0.6,
     xlab = "x", ylab = "y", main = "Eq. (10.7): Simpson's paradox")
abline(lm(y ~ x, data = recs), lty = 2, lwd = 2)
plot((assay1 + assay2)/2, d_ba, pch = 16, cex = 0.6, col = "steelblue",
     xlab = "mean of the two assays", ylab = "difference",
     main = "Eq. (10.8): Bland-Altman")
abline(h = mean(d_ba), lwd = 2)
abline(h = mean(d_ba) + c(-1, 1) * 1.96 * sd(d_ba), col = "red", lty = 2)
hist(rs, breaks = 60, col = "steelblue", border = "white",
     xlab = "pairwise r among INDEPENDENT features",
     main = sprintf("n=%d: |r|>0.5 in %.1f%% of pairs", n_samp, 100*mean(abs(rs)>0.5)))
abline(v = c(-0.5, 0.5), col = "red", lty = 2)
invisible(dev.off())
cat("\nFigure written to", file.path(OUT, "correlation.png"), "\n")

#' ## Decision rules (from `stats.md` Topic 10)
#'
#' 1. Pearson for linear association on a justified scale; Spearman/Kendall for
#'    monotone association or ordinal data.
#' 2. Handle donor structure explicitly (10.7).
#' 3. Use Fisher's z (10.3) for intervals and tests.
#' 4. For method comparison report (10.8)-(10.9), not r.
#' 5. Apply log-ratio transformation before correlating compositions.
#'
#' **Next:** `11_linear_models.R`
